packages/dds/merge-tree/src/partialLengths.ts

```diff
@@ -47,151 +46,172 @@ function latestLEQ(a: PartialSequenceLength[], key: number) {
         const childPartials: PartialSequenceLengths[] = [];
         for (let i = 0; i < block.childCount; i++) {
             const child = block.children[i];
             if (!child.isLeaf()) {
-                const childBlock = child;
+                hasInternalChild = true;
                 if (recur) {
-                    childBlock.partialLengths =
-                        PartialSequenceLengths.combine(mergeTree, childBlock, collabWindow, true);
+                    child.partialLengths =
+                        PartialSequenceLengths.combine(child, collabWindow, true);
                 }
                 // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
-                childPartials.push(childBlock.partialLengths!);
+                childPartials.push(child.partialLengths!);
             }
         }
-        let childPartialsLen = childPartials.length;
-        if (childPartialsLen !== 0) {
-            // Some children are interior nodes
-            if (combinedPartialLengths.partialLengths.length > 0) {
+
+        // If there are no internal children, the PartialSequenceLengths returns from `fromLeaves` is exactly correct.
+        // Otherwise, we must additively combine all of the children partial lengths to get this block's totals.
+        const combinedPartialLengths
+            = hasInternalChild ? new PartialSequenceLengths(collabWindow.minSeq) : leafPartialLengths;
+        if (hasInternalChild) {
+            if (leafPartialLengths.partialLengths.length > 0) {
                 // Some children were leaves; add combined partials from these segments
-                childPartials.push(combinedPartialLengths);
-                childPartialsLen++;
-                combinedPartialLengths = new PartialSequenceLengths(collabWindow.minSeq);
+                childPartials.push(leafPartialLengths);
             }
-            const indices = new Array<number>(childPartialsLen);
-            const childPartialsCounts = new Array<number>(childPartialsLen);
+
+            const childPartialsLen = childPartials.length;
+
+            const sortedChildPartialLengths: PartialSequenceLength[] = [];
             for (let i = 0; i < childPartialsLen; i++) {
-                indices[i] = 0;
-                childPartialsCounts[i] = childPartials[i].partialLengths.length;
-                combinedPartialLengths.minLength += childPartials[i].minLength;
-                combinedPartialLengths.segmentCount += childPartials[i].segmentCount;
+                const { segmentCount, minLength, partialLengths } = childPartials[i];
+                combinedPartialLengths.segmentCount += segmentCount;
+                combinedPartialLengths.minLength += minLength;
+                sortedChildPartialLengths.push(...partialLengths);
             }
-            let outerIndexOfEarliest = 0;
-            let earliestPartialLength: PartialSequenceLength;
-            while (outerIndexOfEarliest >= 0) {
```

Comment:
Not sure if I'm giving too much credit here, but I'd believe it if I was told the old strategy here was chosen for performance reasons. The number of lists we need to search for the minimum of in this is small: capped at size of merge tree's B+ tree branch factor (i.e. 8). It did make the code a lot harder to read, IMO. If we're concerned about this or think it may actually be a hot path, I could restore the logic but stick it in an iterator that would preserve readability. Up to reviewer.

Reply:
you know what, this is bugging me enough that I'm just going to do the iterator thing.

Reply:
latest iteration does that. I'm fine with whatever in this realm.